src/persist-client/src/impl/gc.rs

```diff
@@ -130,72 +130,81 @@ impl GarbageCollector {
         // already gc'd and truncated past new_seqno_since.
 
         let path = req.shard_id.to_string();
-        let state_versions = retry_external(&metrics.retries.external.gc_scan, || async {
-            consensus.scan(&path, SeqNo::minimum()).await
+
+        // pull down the latest version of state, which contains the blobs we want to keep
+        let head = retry_external(&metrics.retries.external.gc_scan, || async {
+            consensus.head(&path).await
         })
-        .await;
-
-        debug!(
-            "gc {} for [{},{}) got {} versions from scan",
-            req.shard_id,
-            req.old_seqno_since,
-            req.new_seqno_since,
-            state_versions.len()
-        );
-
-        // It'd be minor-ly more efficient to reverse the order and first build
-        // up non_deleteable_blobs, and then check all state_versions with seqno
-        // < new_seqno_since so our hashmap grows to O(non_deleteable) rather
-        // than O(deleteable+non_deleteable). This version of the code reads
-        // slightly more obviously, so hold off on that until/if we see it be a
-        // problem in practice.
-        let mut deleteable_blobs = HashSet::new();
-        for state_version in state_versions {
-            if state_version.seqno < req.new_seqno_since {
-                Self::for_all_keys(&state_version, |key| {
-                    // It's okay (expected) if the key already exists in
-                    // deleteable_blobs, it may have been present in previous
-                    // versions of state.
-                    deleteable_blobs.insert(key.to_owned());
-                });
-            } else if state_version.seqno == req.new_seqno_since {
-                Self::for_all_keys(&state_version, |key| {
-                    // It's okay (expected) if the key doesn't exist in
-                    // deleteable_blobs, it may have been added in this version
-                    // of state.
-                    let _ = deleteable_blobs.remove(key);
-                });
-            } else {
-                // Sanity check the loop logic.
-                assert!(state_version.seqno > req.new_seqno_since);
-                break;
+        .await
+        .expect("cannot have been asked to gc a shard without any states");
+
+        let mut required_blobs = HashSet::new();
+        Self::for_all_keys(&head, |key| {
+            let _ = required_blobs.insert(key.to_owned());
+        });
+
+        let mut gc_seqno_progress = SeqNo::minimum();
+        while gc_seqno_progress < head.seqno {
+            let state_versions = retry_external(&metrics.retries.external.gc_scan, || async {
+                consensus.scan(&path, gc_seqno_progress, 20).await
```

Comment:
nit: make this 20 at least a named const somewhere (even in this file or fn is fine). I try to avoid having hardcoded numbers anywhere inline in code